### PR TITLE
Update wandb_callbacks.py

### PR DESCRIPTION
**Pull request by zhengyu-yang**
UploadCodeAsArtifact: Add the functionality to use git to decide which files are source files. Can upload all files that are not ignored by git instead of all '*.py' file.
UploadCodeAsArtifact: Add @rank_zero_only
UploadCheckpointsAsArtifact: Add @rank_zero_only
UploadCheckpointsAsArtifact: ckpts are output of the run, so use `experiment.log_artifact(ckpts)` instead

### CODE REVIEW

**Comment by ashleve** [comment]
@zhengyu-yang Hi there,
Thanks for the contribution!

Good idea with rank zero and `experiment.log_artifact`

I run `python run.py callbacks=wandb logger=wandb` and got the following error:
```
  File "/home/ash/Desktop/lightning-hydra-template/lightning-hydra-template/src/callbacks/wandb_callbacks.py", line 70, in on_train_start
    and (not path.is_relative_to(git_dir_path))  # ignore files in .git
AttributeError: 'PosixPath' object has no attribute 'is_relative_to'
```
I believe you might be using python 3.9? I tried running it on 3.8
The template is supposed to be compatible with 3.7, 3.8 and 3.9

**Comment by zhengyu-yang** [comment]
Yes, I am running on 3.9. I will check if there is a better way.

**Comment by zhengyu-yang** [comment]
I have updated it for compatibility

**Comment by zhengyu-yang** [comment]
@ashleve Can you have a look at this?

**Comment by ashleve** [comment]
@zhengyu-yang Sorry for late response, I kind of missed the update of this PR.

The UploadCodeAsArtifact improvement is very cool, would be happy to merge this. Let me propose a couple improvements

**Comment by ashleve** [comment]
The template uses black and flake8, please run them to format the code:
```bash
pip install pre-commit
pre-commit install
pre-commit run -a
```

File: src/callbacks/wandb_callbacks.py [hunk 1]
```diff
@@ -41,18 +51,38 @@ def on_train_start(self, trainer, pl_module):
...
-            code.add_file(path)
+
+        if self.use_git:
+            # get .git folder
+            # https://alexwlchan.net/2020/11/a-python-function-to-ignore-a-path-with-git-info-exclude/
+            git_dir_path = Path(subprocess.check_output(["git", "rev-parse", "--git-dir"]).strip().decode("utf8")).resolve()
+
+            for path in Path(self.code_dir).rglob('*'):
+                if (path.is_file()
+                        and (not is_relative_to(path, git_dir_path))  # ignore files in .git
```
Comment (ashleve) [inline src/callbacks/wandb_callbacks.py:79]: I replaced this line with:
```python
and not git_dir_path in path.parents
```
and it seems to still ignore `.git/` correctly (this way we don't need the `is_relative_to` method?)
Is this correct or am I missing something?

File: src/callbacks/wandb_callbacks.py [hunk 2]
```diff
@@ -41,18 +51,38 @@ def on_train_start(self, trainer, pl_module):
...
+                        and (subprocess.run(['git', 'check-ignore', '-q', str(path)]).returncode == 1)):  # ignore files ignored by git
+                    code.add_file(str(path), name=str(path.relative_to(self.code_dir)))
+
+        else:
+            for path in Path(self.code_dir).rglob('*.py'):
+                code.add_file(str(path), name=str(path.relative_to(self.code_dir)))
 
         experiment.use_artifact(code)
```
Comment (ashleve) [inline src/callbacks/wandb_callbacks.py:87]: For some reason the run view doesn't show me its code artifact when using `use_artifact` but it works correctly with `log_artifact`. Although it always shows up correctly in global view.
Do you also have this issue?
Maybe this callback also should be using `log_artifact` for now

File: src/callbacks/wandb_callbacks.py [hunk 3]
```diff
@@ -41,18 +51,38 @@ def on_train_start(self, trainer, pl_module):
...
 
 class UploadCodeAsArtifact(Callback):
-    """Upload all *.py files to wandb as an artifact, at the beginning of the run."""
+    """Upload all code files to wandb as an artifact, at the beginning of the run."""
+
+    def __init__(self, code_dir: str, use_git: bool = True):
+        """
 
-    def __init__(self, code_dir: str):
+        :param code_dir:
```
Comment (ashleve) [inline src/callbacks/wandb_callbacks.py:59]: Please follow the google docstring convention as it's already used in `utils.py`

**Review by ashleve — CHANGES_REQUESTED** [review]

File: src/callbacks/wandb_callbacks.py [hunk 4]
```diff
@@ -41,18 +51,38 @@ def on_train_start(self, trainer, pl_module):
 
 
 class UploadCodeAsArtifact(Callback):
-    """Upload all *.py files to wandb as an artifact, at the beginning of the run."""
+    """Upload all code files to wandb as an artifact, at the beginning of the run."""
+
+    def __init__(self, code_dir: str, use_git: bool = True):
```
Comment (ashleve) [inline src/callbacks/wandb_callbacks.py:56]:
```suggestion
    def __init__(self, code_dir: str, use_gitignore: bool = True):
```

File: src/callbacks/wandb_callbacks.py (hunk 1 above)
Comment (ashleve) [reply src/callbacks/wandb_callbacks.py:79]: Actually this should be faster:
```python
and not str(git_dir_path) in str(path)
```

File: src/callbacks/wandb_callbacks.py (hunk 2 above)
Comment (zhengyu-yang) [reply src/callbacks/wandb_callbacks.py:87]: In my experience, sometimes the artifacts do not show up during the run but will eventually show up after the run is finished

File: src/callbacks/wandb_callbacks.py (hunk 1 above)
Comment (zhengyu-yang) [reply src/callbacks/wandb_callbacks.py:79]: I think `str(git_dir_path) in str(path)` is potentially problematic in some edge cases. For example:
```python
# example .git directory
git_dir_path = Path('/home/foo/repo/.git')
# a path that is actually not in the example .git directory
path = Path('/home/foo/repo/home/foo/repo/.git/bar.py')
```

File: src/callbacks/wandb_callbacks.py (hunk 3 above)
Comment (zhengyu-yang) [reply src/callbacks/wandb_callbacks.py]: This is solved

File: src/callbacks/wandb_callbacks.py (hunk 1 above)
Comment (zhengyu-yang) [reply src/callbacks/wandb_callbacks.py]: Currently, my `is_relative_to` directly copies the python 3.9 implementation of `is_relative_to`, which I believe is the safest practice.

**Comment by zhengyu-yang** [comment]
@ashleve I have updated code the according to your suggestions. For the two remaining change requests, I am not sure if they are still necessary.

File: src/callbacks/wandb_callbacks.py (hunk 1 above)
Comment (ashleve) [reply src/callbacks/wandb_callbacks.py]: How about:
```python
and not str(path).startswith(str(git_dir_path))
```
Seems like covers all edge cases?

File: src/callbacks/wandb_callbacks.py [hunk 5]
```diff
@@ -67,6 +106,10 @@ def __init__(self, ckpt_dir: str = "checkpoints/", upload_best_only: bool = Fals
         self.ckpt_dir = ckpt_dir
         self.upload_best_only = upload_best_only
 
+    @rank_zero_only
+    def on_keyboard_interrupt(self, trainer, pl_module):
+        self.on_train_end(trainer, pl_module)
+
```
Comment (ashleve) [inline src/callbacks/wandb_callbacks.py:112]: I added this hook so ckpts will still be uploaded even when run is interrupted with `ctrl + c`

File: src/callbacks/wandb_callbacks.py (hunk 1 above)
Comment (zhengyu-yang) [reply src/callbacks/wandb_callbacks.py]: Solved. However, I just want to say that my personal philosophy is that if a functionality has a standard library implementation, we should try to follow that.

Comment (zhengyu-yang) [reply src/callbacks/wandb_callbacks.py]: @ashleve Please have a look 😄

Comment (ashleve) [reply src/callbacks/wandb_callbacks.py]: Thanks, I like your approach!
It's just that I believe this specific repo will be the most useful when it's as simple and clean as possible. I'm afraid that after taking many such design choices the code could just get too complex

**Comment by ashleve** [comment]
@zhengyu-yang Merged. Thanks again, great PR!

**Pull request merged**